July news (#803)
ropensci/roweb3: content/blog/2024-07-19-news-july-2024/index.Rmd

### Last call: your opinion on the CRAN submission process!

Shared by Lluís Revilla and Heather Turner in our Slack workspace, a crucial survey ending today!

>  "If you have R package development experience and would like to share your thoughts on the CRAN submission process, please fill this short survey from the CRAN Cookbook project!"

Please find the [Google form](https://forms.gle/umdew9KHWeQSehq8A) and read more about the exciting cookbook project in this [post by Jasmine D.](https://www.linkedin.com/pulse/improving-skills-r-package-maintainers-support-from-consortium-daly-tlt4e).

### Robust type-checking with r-lib

Don't miss this insightful short post by Josiah Parry, ["Type safe(r) R code"](https://josiahparry.com/posts/2024-06-30-type-safety/).
A related older blog post is ["Checking the inputs of your R functions"](https://blog.r-hub.io/2022/03/10/input-checking/) by Hugo Gruson, Sam Abbott, Carl Pearson.

### The one with all the useR! links

The useR! 2024 conference featured quite a few talks relevant to package development, beside the talks we mentioned in the [HQ section](#resources-from-the-ropensci-community-at-user-2024).

> Not all recordings are available yet, but make sure to check out the useR! YouTube channel.

Reviewer: ```suggestion
Not all recordings are available yet, but make sure to check out the [useR! YouTube channel](https://www.youtube.com/@useRConference_global).
```

If we missed any relevant content, please [get in touch](https://github.com/ropensci-org/monthly/issues) so we might add missing pieces to our next newsletter!

#### Retrospectives

> Kurt Hornik gave a keynote talks on [More than 25 years of CRAN](https://userconf2024.sched.com/event/1c9Jm/keynote-more-than-25-years-of-cran-kurt-hornik-wirtschaftsuniversitat-wien?iframe=no&w=100%&sidebar=yes&bg=no) ([slides](https://static.sched.com/hosted_files/userconf2024/41/user2024_CRAN.pdf)) and Torsten Hothorn on [Some things you can’t read from

Reviewer: ```suggestion
Kurt Hornik and Torsten Hothorn gave keynote talks ["More than 25 years of CRAN"](https://userconf2024.sched.com/event/1c9Jm/keynote-more-than-25-years-of-cran-kurt-hornik-wirtschaftsuniversitat-wien?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/41/user2024_CRAN.pdf)) and ["Some things you can’t read from
```

> a NEWS file](https://userconf2024.sched.com/event/1c9JC/keynote-some-things-you-cant-read-from-a-news-file-torsten-hothorn-university-of-zurich?iframe=yes&w=100%&sidebar=yes&bg=no) ([slides](https://static.sched.com/hosted_files/userconf2024/b6/useR_2024_TH%20%282%29.pdf)) about maintaining a package for decades.

Reviewer: ```suggestion
a NEWS file"](https://userconf2024.sched.com/event/1c9JC/keynote-some-things-you-cant-read-from-a-news-file-torsten-hothorn-university-of-zurich?iframe=yes&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/b6/useR_2024_TH%20%282%29.pdf)) about maintaining a package for decades.
```

> Edzer Pebesma and Roger Bivand reported on [The Retirement of R Packages with Many Reverse Dependencies](https://userconf2024.sched.com/event/1c90V/r-evolution-the-retirement-of-r-packages-with-many-reverse-dependencies-edzer-pebesma-university-of-muenster-roger-bivand-norwegian-school-of-economics?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/2a/UseR2024.pdf).

Reviewer: ```suggestion
Edzer Pebesma and Roger Bivand reported on ["The Retirement of R Packages with Many Reverse Dependencies"](https://userconf2024.sched.com/event/1c90V/r-evolution-the-retirement-of-r-packages-with-many-reverse-dependencies-edzer-pebesma-university-of-muenster-roger-bivand-norwegian-school-of-economics?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/2a/UseR2024.pdf)).
```

#### On validation of R packages

> Coline Zeballos and Yann Féhat from the R Validation Hub discussed [how to support (pharma) companies with validation of R Packages](https://userconf2024.sched.com/event/1dyXW/supporting-companies-with-validation-of-r-packages-yann-feat-mainanalytics-gmbh-coline-zeballos-r-validation-hub?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/4a/Supporting%20Companies%20With%20Validation%20of%20R%20Packages_%20A%20Regulatory%20Repository.pdf). They use a toolset based on [r-hub/repos](https://github.com/r-hub/repos) and the [riskmetric](https://pharmar.github.io/riskmetric/) package.

Reviewer: ```suggestion
Coline Zeballos and Yann Féhat from the R Validation Hub discussed [how to support (pharma) companies with validation of R Packages](https://userconf2024.sched.com/event/1dyXW/supporting-companies-with-validation-of-r-packages-yann-feat-mainanalytics-gmbh-coline-zeballos-r-validation-hub?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/4a/Supporting%20Companies%20With%20Validation%20of%20R%20Packages_%20A%20Regulatory%20Repository.pdf)). They use a toolset based on [r-hub/repos](https://github.com/r-hub/repos) and the [riskmetric](https://pharmar.github.io/riskmetric/) package.
```

Szymon Maksymiuk and Lorenzo Braschi presented a [Deep Dive Into Industry R

> Package Quality Assessment](https://userconf2024.sched.com/event/1c8z7/deep-dive-into-industry-r-package-quality-assessment-szymon-maksymiuk-lorenzo-braschi-roche?iframe=no). Beside introducing the concepts, they mentioned three open-source R packages that they created: [checked](https://github.com/Genentech/checked) for running reverse dependencies checks; [covtracer](https://genentech.github.io/covtracer/) for contextualizing tests, built using covr test traces; [rd2markdown](https://github.com/Genentech/rd2markdown) for converting .Rd files into Markdown.

Reviewer: ```suggestion
Package Quality Assessment](https://userconf2024.sched.com/event/1c8z7/deep-dive-into-industry-r-package-quality-assessment-szymon-maksymiuk-lorenzo-braschi-roche?iframe=no). Beside introducing the concepts, they mentioned three open-source R packages that they created: [checked](https://github.com/Genentech/checked) for running reverse dependencies checks; [covtracer](https://genentech.github.io/covtracer/) for contextualizing tests using covr test traces; [rd2markdown](https://github.com/Genentech/rd2markdown) for converting .Rd files into Markdown.
```

Also on reverse dependencies checks, Pawel Rucki and André Veríssimo presented

>  [{verdepcheck} - A Tool for Dependencies Check](https://userconf2024.sched.com/event/1cC1c/verdepcheck-a-tool-for-dependencies-check-pawel-rucki-andre-verissimo-roche?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/2d/%7Bverdepcheck%7D%20-%20A%20Tool%20for%20Dependencies%20Check.pdf), [verdepcheck](https://insightsengineering.github.io/verdepcheck/main/).

Reviewer: ```suggestion
 [{verdepcheck} - A Tool for Dependencies Check](https://userconf2024.sched.com/event/1cC1c/verdepcheck-a-tool-for-dependencies-check-pawel-rucki-andre-verissimo-roche?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/2d/%7Bverdepcheck%7D%20-%20A%20Tool%20for%20Dependencies%20Check.pdf) | [Package Docs](https://insightsengineering.github.io/verdepcheck/main/)).
```

> Franciszek Walkowiak's topic was [Systems Integration Tests for R Package Cohorts](https://userconf2024.sched.com/event/1c8yp/systems-integration-tests-for-r-package-cohorts-franciszek-walkowiak-roche?iframe=no&w=100%&sidebar=yes&bg=no), including the introductions to two open-source utilities, [scribe](https://github.com/insightsengineering/scribe) that creates complete build, check and install reports for a collection of R projects and [locksmith](https://github.com/insightsengineering/locksmith) that helps with `renv.lock` creation. [Slides](https://static.sched.com/hosted_files/userconf2024/1c/systems-integration-tests-for-r-package-cohorts.pdf).

Reviewer: ```suggestion
Franciszek Walkowiak discussed [Systems Integration Tests for R Package Cohorts](https://userconf2024.sched.com/event/1c8yp/systems-integration-tests-for-r-package-cohorts-franciszek-walkowiak-roche?iframe=no&w=100%&sidebar=yes&bg=no), including the introductions to two open-source utilities, [scribe](https://github.com/insightsengineering/scribe) that creates complete build, check and install reports for a collection of R projects and [locksmith](https://github.com/insightsengineering/locksmith) that helps with `renv.lock` creation ([Slides](https://static.sched.com/hosted_files/userconf2024/1c/systems-integration-tests-for-r-package-cohorts.pdf)).
```

#### On good practice

> Daniel Sabanés Bové introduced openstatsware's work on _minimal viable_ good practice standards for R packages, in an [online talk](https://virtualuser2024.sched.com/event/1eBda/openstatsguide-minimum-viable-good-practices-for-high-quality-statistical-software-packages-daniel-sabanes-bove-rconis?iframe=no&w=&sidebar=yes&bg=no).

Reviewer: ```suggestion
Daniel Sabanés Bové introduced openstatsware's work on [_minimal viable_ good practice standards for R packages](https://virtualuser2024.sched.com/event/1eBda/openstatsguide-minimum-viable-good-practices-for-high-quality-statistical-software-packages-daniel-sabanes-bove-rconis?iframe=no&w=&sidebar=yes&bg=no).
```

Pedro Silva listed [Seven Deadly Sins Holding You

> Back as a Software Developer](https://userconf2024.sched.com/event/1c8zm/seven-deadly-sins-holding-you-back-as-a-software-developer-pedro-silva?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/86/Seven%20Deadly%20Sins%20Holding%20You%20Back%20as%20a%20Software%20Developer.pdf).

Reviewer: ```suggestion
Back as a Software Developer](https://userconf2024.sched.com/event/1c8zm/seven-deadly-sins-holding-you-back-as-a-software-developer-pedro-silva?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/86/Seven%20Deadly%20Sins%20Holding%20You%20Back%20as%20a%20Software%20Developer.pdf)).
```

> Hugo Gruson had a poster on [A reproducible analysis of CRAN Task Views to understand the state of an R package ecosystem](https://userconf2024.sched.com/event/1cDry/poster-session-guided-tours-a-reproducible-analysis-of-cran-task-views-to-understand-the-state-of-an-r-package-ecosystem-hugo-gruson-dataorg?iframe=no&w=100%&sidebar=yes&bg=no). Find the [live analysis](https://epiverse-connect.github.io/ctv-analysis/).

Reviewer: ```suggestion
Hugo Gruson had a poster on [A reproducible analysis of CRAN Task Views to understand the state of an R package ecosystem](https://userconf2024.sched.com/event/1cDry/poster-session-guided-tours-a-reproducible-analysis-of-cran-task-views-to-understand-the-state-of-an-r-package-ecosystem-hugo-gruson-dataorg?iframe=no&w=100%&sidebar=yes&bg=no). See the [live analysis](https://epiverse-connect.github.io/ctv-analysis/).
```

#### On package design

> Hugo Gruson highlighted the benefits of using S3 classes for [interoperability in Existing Software Ecosystems](https://userconf2024.sched.com/event/1c8z3/building-interoperability-in-existing-software-ecosystems-with-s3-classes-hugo-gruson-dataorg?iframe=no). [Slides](https://static.sched.com/hosted_files/userconf2024/ff/2024-07-09_Gruson%20-%20Building%20Interoperability%20with%20S3%20Classes.pdf).

Reviewer: ```suggestion
Hugo Gruson highlighted the benefits of using S3 classes for [interoperability in Existing Software Ecosystems](https://userconf2024.sched.com/event/1c8z3/building-interoperability-in-existing-software-ecosystems-with-s3-classes-hugo-gruson-dataorg?iframe=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/ff/2024-07-09_Gruson%20-%20Building%20Interoperability%20with%20S3%20Classes.pdf)).
```

See also his recent [blog posts on the topic](https://epiverse-trace.github.io/blog.html#category=S3).

Ligia Adamska used an onion analogy to explain [Layered Design for R Package
Development: Meeting the Needs of

> Pharmaceutical R&D Stakeholders](https://userconf2024.sched.com/event/1c8z0/layered-design-for-r-package-development-meeting-the-needs-of-pharmaceutical-rd-stakeholders-ligia-adamska-msd-switzerland?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/7a/UseR_presentation_Final.pdf).

Reviewer: ```suggestion
Pharmaceutical R&D Stakeholders](https://userconf2024.sched.com/event/1c8z0/layered-design-for-r-package-development-meeting-the-needs-of-pharmaceutical-rd-stakeholders-ligia-adamska-msd-switzerland?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/7a/UseR_presentation_Final.pdf)).
```

#### On tools

Daphne Grasselly, Franciszek Walkowiak and Pawel Rucki lead a tutorial on [Streamlining R package
development with GitHub Actions

> Workflows](https://userconf2024.sched.com/event/1c8yI/tutorial-streamlining-r-package-development-with-github-actions-workflows-daphne-grasselly-pawel-rucki-roche-dinakar-kulkarni-genentech-pre-registration-required?iframe=no) whose [slides](https://static.sched.com/hosted_files/userconf2024/19/streamlining-r-package-development-with-github-actions-workflows.pdf) are available.

Reviewer: ```suggestion
Workflows](https://userconf2024.sched.com/event/1c8yI/tutorial-streamlining-r-package-development-with-github-actions-workflows-daphne-grasselly-pawel-rucki-roche-dinakar-kulkarni-genentech-pre-registration-required?iframe=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/19/streamlining-r-package-development-with-github-actions-workflows.pdf)).
```

Emil Hvitfeldt explained how to [make better error messages with rlang and cli](https://virtualuser2024.sched.com/event/1eBdO/making-better-error-messages-with-rlang-and-cli-emil-hvitfeldt-posit-pbc?iframe=no&w=&sidebar=yes&bg=no).

> Ella Kaye shared her insights on [C for R users](https://userconf2024.sched.com/event/1c8zS/c-for-r-users-ella-kaye-university-of-warwick?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/84/c-for-r-users.pdf).

Reviewer: ```suggestion
Ella Kaye shared her insights on [C for R users](https://userconf2024.sched.com/event/1c8zS/c-for-r-users-ella-kaye-university-of-warwick?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/84/c-for-r-users.pdf)).
```

> Davis Vaughan introduced [tree-sitter, an efficient incremental parsing library](https://userconf2024.sched.com/event/1c90k/improving-development-tooling-with-an-r-grammar-for-tree-sitter-davis-vaughan-posit?iframe=no&w=100%&sidebar=yes&bg=no). [R package with bindings to tree-sitter](https://davisvaughan.github.io/r-tree-sitter/) whose README states "tree-sitter is useful for a number of things, including syntax highlighting, go-to definition, code reshaping, and more."

Reviewer: ```suggestion
Davis Vaughan introduced [tree-sitter, an efficient incremental parsing library](https://userconf2024.sched.com/event/1c90k/improving-development-tooling-with-an-r-grammar-for-tree-sitter-davis-vaughan-posit?iframe=no&w=100%&sidebar=yes&bg=no) and the R package [treesitter, which provides bindings to tree-sitter](https://davisvaughan.github.io/r-tree-sitter/) whose README states "tree-sitter is useful for a number of things, including syntax highlighting, go-to definition, code reshaping, and more."
```

#### On multilingualism

Elio Campitelli spoke about [Building Bilingual Bridges
with Multilingual Manuals](https://userconf2024.sched.com/event/1c8yj/building-bilingual-bridges-with-multilingual-manuals-elio-campitelli-universidad-de-buenos-aires?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/3d/UseR2024-1.pdf). See also their [post on our blog](https://ropensci.org/blog/2024/07/09/multilingual-documentation-wg/)!

#### On debugging

> Shannon Pileggi delivered a tutorial on [debugging in R](https://events.linuxfoundation.org/user/program/schedule/). [Materials](https://github.com/rstats-wtf/wtf-debugging).

Reviewer: ```suggestion
Shannon Pileggi delivered a tutorial on [debugging in R](https://events.linuxfoundation.org/user/program/schedule/) ([Materials](https://github.com/rstats-wtf/wtf-debugging)).
```

> Antoine Fabri gave an overview of the motivations behind, and features of, his [constructive package](https://userconf2024.sched.com/event/1c8ym/constructive-a-nicer-dput-using-idiomatic-constructors-antoine-fabri-cynkra?iframe=no&w=100%&sidebar=yes&bg=no), that among other things can be useful for troubleshooting. [constructive](https://cynkra.github.io/constructive/).

Reviewer: ```suggestion
Antoine Fabri gave an [overview of the motivations behind, and features of, his constructive package](https://userconf2024.sched.com/event/1c8ym/constructive-a-nicer-dput-using-idiomatic-constructors-antoine-fabri-cynkra?iframe=no&w=100%&sidebar=yes&bg=no), which, among other things, can be useful for troubleshooting ([Package docs](https://cynkra.github.io/constructive/)).
```

#### On wrapping APIs

> Hadley Wickham introduced and demo-ed his [httr2 package](https://userconf2024.sched.com/event/1c8zP/web-apis-with-httr2-hadley-wickham-posit?iframe=no&w=100%&sidebar=yes&bg=no). [httr2](https://httr2.r-lib.org/).

Reviewer: ```suggestion
Hadley Wickham introduced and demo-ed his [httr2 package](https://userconf2024.sched.com/event/1c8zP/web-apis-with-httr2-hadley-wickham-posit?iframe=no&w=100%&sidebar=yes&bg=no) ([Package docs](https://httr2.r-lib.org/)).
```

Simon Haller explained the [Automated Generation

> of R Client Packages for RESTful APIs](https://userconf2024.sched.com/event/1c8ys/automated-generation-of-r-client-libraries-for-restful-apis-using-openapi-specification-and-the-open-simon-haller-raiffeisen?iframe=no&w=100%&sidebar=yes&bg=no). [Slides](https://static.sched.com/hosted_files/userconf2024/87/User_R_openapi_generator_simonhaller.pdf). See also Jon Harmon's [work on the same topic](https://beekeeper.api2r.org/).

Reviewer: ```suggestion
of R Client Packages for RESTful APIs](https://userconf2024.sched.com/event/1c8ys/automated-generation-of-r-client-libraries-for-restful-apis-using-openapi-specification-and-the-open-simon-haller-raiffeisen?iframe=no&w=100%&sidebar=yes&bg=no) ([Slides](https://static.sched.com/hosted_files/userconf2024/87/User_R_openapi_generator_simonhaller.pdf)). See also Jon Harmon's [work on the same topic](https://beekeeper.api2r.org/).
```

#### On a last resort for archived CRAN packages

> Henrik Bengtsson and Lluís Revilla had a poster about their [CRANhaven project](https://www.cranhaven.org/) that is a backup solution for end-users when a package falls of CRAN (and that is built using R-universe!).

Reviewer: ```suggestion
Henrik Bengtsson and Lluís Revilla had a poster about their [CRANhaven project](https://www.cranhaven.org/), a backup solution for end-users when a package falls of CRAN (and which is built using R-universe!).
```
